Sources/GRPC/AsyncAwaitSupport/GRPCAsyncResponseStreamWriter.swift

```diff
@@ -19,26 +19,90 @@
 /// Writer for server-streaming RPC handlers to provide responses.
 @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
 public struct GRPCAsyncResponseStreamWriter<Response: Sendable>: Sendable {
+  public struct RecordingAsyncSequence: AsyncSequence {
+    public typealias Element = (Response, Compression)
+
+    private let stream: AsyncStream<(Response, Compression)>
+
+    init(stream: AsyncStream<(Response, Compression)>) {
+      self.stream = stream
+    }
+
+    public func makeAsyncIterator() -> AsyncIterator {
+      AsyncIterator(iterator: self.stream.makeAsyncIterator())
+    }
+
+    public struct AsyncIterator: AsyncIteratorProtocol {
+      private var iterator: AsyncStream<(Response, Compression)>.AsyncIterator
+
+      init(iterator: AsyncStream<(Response, Compression)>.AsyncIterator) {
+        self.iterator = iterator
+      }
+
+      public mutating func next() async -> Element? {
+        await self.iterator.next()
+      }
+    }
+  }
+
+  @usableFromInline
+  enum Backing: Sendable {
+    case asyncWriter(AsyncWriter<Delegate>)
+    case closure(@Sendable ((Response, Compression)) async -> Void)
+  }
+
   @usableFromInline
   internal typealias Element = (Response, Compression)
 
   @usableFromInline
   internal typealias Delegate = AsyncResponseStreamWriterDelegate<Response>
 
   @usableFromInline
-  internal let asyncWriter: AsyncWriter<Delegate>
+  internal let backing: Backing
 
   @inlinable
   internal init(wrapping asyncWriter: AsyncWriter<Delegate>) {
-    self.asyncWriter = asyncWriter
+    self.backing = .asyncWriter(asyncWriter)
+  }
+
+  @inlinable
+  internal init(onWrite: @escaping @Sendable ((Response, Compression)) async -> Void) {
+    self.backing = .closure(onWrite)
   }
 
   @inlinable
   public func send(
     _ response: Response,
     compression: Compression = .deferToCallDefault
   ) async throws {
-    try await self.asyncWriter.write((response, compression))
+    switch self.backing {
+    case let .asyncWriter(writer):
+      try await writer.write((response, compression))
+
+    case let .closure(closure):
+      await closure((response, compression))
+    }
+  }
+
+  public struct RecordingWriter {
+    /// The actual writer.
+    public let writer: GRPCAsyncResponseStreamWriter<Response>
+    /// An `AsyncSequence` with the written responses and their compression.
+    public let responses: RecordingAsyncSequence
+  }
+
+  /// Creates a new `GRPCAsyncResponseStreamWriter` backed by an ``RecordingAsyncSequence``.
+  /// This is mostly useful for testing purposes where one wants to observe the output of a method invocation.
+  public static func makeRecordingWriter() -> RecordingWriter {
+    var continuation: AsyncStream<(Response, Compression)>.Continuation!
+    let asyncStream = AsyncStream<(Response, Compression)> { cont in
+      continuation = cont
+    }
+    let writer = Self.init { [continuation] in
```

Comment:
There's an awkward wrinkle in this code: we never finish the async stream. gRPC does this internally on the async writer (IIRC) after the user method exits. We need to rework this to do it here too.

Reply:
I don't know if I fully agree on this. The only way to really finish the sequence of writes would be to to __move__ the writer into the user function and implement a deinit on the writer to finish the sequence. However, I don't think we can spell this right now.
Furthermore, I don't think we necessarily need to finish the sequence. In your tests, you should be able to use `prefix()` and `collect()` to assert how many writes have been issued.
Open to other ideas how you think we should finish this sequence!

Reply:
My concern is that it's non-obvious that without using `prefix()` or collecting a fixed number of messages that iterating over the sequence will never terminate. Not being able to safely iterate an async sequence is a serious API flaw IMHO.

---

With the end goal of making services testable I think we may need to look at a bigger picture, bearing in mind we also need to provide a context to the caller.

I wonder if we can do something like providing a service wrapper/context/thing which can provide the right "bits" to the caller and also deal with finishing the stream. I'd want to play around with ideas some more to flesh this out but it could be something like:

```swift
struct TestingContext {
    var logger: Logger

    init(logger: Logger) {
        self.logger = logger
    }

    func withRPCContext<Request: Sendable, Response: Sendable>(
        execute function: (Request, GRPCAsyncServerCallContext) async throws -> Response
    ) async throws -> Response {
        fatalError()
    }

    func withRPCContext<Request: Sendable, Response: Sendable>(
        requestType: Request.Type,
        responseType: Response.Type,
        execute function: (GRPCAsyncRequestStream<Request>, GRPCAsyncServerCallContext) async throws -> Response,
        writingRequests: (WrappedAsyncStreamContinuation<Request>) -> Void
    ) async throws -> Response {
        fatalError()
    }

    func withRPCContext<Request: Sendable, Response: Sendable>(
        request: Request,
        responseType: Response.Type,
        execute function: (Request, GRPCAsyncResponseStreamWriter<Response>.AsyncSequence, GRPCAsyncServerCallContext) async throws -> Void
    ) async throws -> GRPCAsyncResponseStreamWriter<Response> {
        fatalError()
    }

    func withRPCContext<Request: Sendable, Response: Sendable>(
        requestType: Request.Type,
        responseType: Response.Type,
        execute function: (GRPCAsyncRequestStream<Request>, GRPCAsyncResponseStreamWriter<Response>..AsyncSequence, GRPCAsyncServerCallContext) async throws -> Void,
        readingAndWriting: (WrappedAsyncStreamContinuation<Request>, GRPCAsyncResponseStreamWriter<Response>) async throws -> Void
    ) async throws {
        fatalError()
    }
}
```

The idea being to have it execute the function and provide the tester with the appropriate pieces to safely test the function. This should also include things like setting request headers and so on (although I haven't included that here). Rough example:

```swift
let context = TestingContext(logger: Logger(label: "foo"))
let echo = EchoAsyncProvider()

let echoResponse = try await context.withRPCContext(
    request: .with { $0.text = "hello" },
    execute: echo.get(request:context:)
)

try await context.withRPCContext(
    requestType: Echo_EchoRequest.self,
    responseType: Echo_EchoResponse.self,
    execute: echo.update(requestStream:responseStream:context:)
) { requests, responses in
    try await withThrowingTaskGroup(of: Void.self) { group in
        group.addTask {
            requests.write(...)
        }

        group.addTask {
            for try await response in responses {
                print(response)
            }
        }

        try await group.waitForAll()
    }
}
```

Reply:
While I fully agree with what you are laying out here, I think I am working on a slightly different but related corner of the testability. I have misphrased my motivation above a bit. My goal is not to test a method implementation per-se which I agree we need a more full blown solution like you pointed out since we need to solve the context and more.

However, my goal is to make both the `GRPCAsyncRequestStream` and the `GRPCAsyncResponseWriter` testable. You can imagine a scenario where you take the writer and pass it further along in your system and you wanna make these methods testable. However, due to the fact that the writer is not testable as of this moment it becomes impossible.

I see two ways out here, either we make the writer testable i.e. by surfacing a never ending `AsyncSequence` from it. Or we say this type is not testable and we require users to re-wrap the writer if they wanna pass it along and make it testable. The latter seems somewhat wrong to me since we would force a lot of adopters to write exactly the same code for testability sake.

On another note, I think that the writer test `AsyncSequence` is not ending is totally fine since the API that the writer exposes is just `write(_:)` so it _must_ produce an infinite sequence. Infinite sequences are quite normal e.g. `CurrentValueSubject<Int, Never>` from Combine. Let me know what you think!